File: 0316-Remove Duplicate Letters/0316-Remove Duplicate Letters.cpp
class Solution {
public:
    string removeDuplicateLetters(string s) {
        vector<int> count(26);
        for (char c : s)
        {
            ++count[c - 'a'];
        }
        
        vector<bool> visited(26);
        string result;
        for (char c : s)
        {
            --count[c - 'a'];
            if (visited[c - 'a'])
            {
                continue;
            }
            
            while (result.size() && c < result.back() && count[result.back() - 'a'])
            {
                visited[result.back() - 'a'] = false;
                result.pop_back();
            }
            
            visited[c - 'a'] = true;
            result.push_back(c);
        }
        
        return result;
    }
};

File: 0301-0400/0316-Remove Duplicate Letters/0316-Remove Duplicate Letters.py
import collections
class Solution:
    def removeDuplicateLetters(self, s: str) -> str:
        table = collections.Counter(s)
        visited = set()
        result = []
        for c in s:
            table[c] -= 1
            if c in visited:
                continue
                
            while result and c < result[-1] and table[result[-1]] > 0:
                visited.remove(result[-1])
                result.pop()
                
            result.append(c)
            visited.add(c)
            
        return ''.join(result)
